main.nf

```diff
@@ -1,65 +1,96 @@
 #!/usr/bin/env nextflow 
 
-// Workflow to perform k-fold cross validation over a downloaded dataset (via accession id) or an input dataset
+// Workflow to perform k-fold cross validation over FTP downloaded datasets 
+// as specified in the csv file in 'data/datasets.txt' 
 
-// download query data
+// download data
 if(params.download_data.run == "True"){
-    query_data = params.download_data.query_output_dir
-    query_n_clust = params.download_data.query_num_clust.toString()
-    query_markers = "marker_genes_" + query_n_clust + ".tsv"
-
+	
+	// initialise required channels 
+	DATASET_IDS = Channel.create()
+	NUM_CLUST = Channel.create()
+	BARCODE_COLUMN = Channel.create()
+	CELL_LABEL_COLUMN = Channel.create()
+	
+	// extract matrix types required by the tools, conditioned on them being 'on' 
+	tool_switch = ["True":0, "False":1]
+	garnett_matrix_type = [params.garnett.matrix_type, null][tool_switch[params.garnett.run]]
+	scmap_cluster_matrix_type = [params.scmap_cluster.matrix_type, null][tool_switch[params.scmap_cluster.run]]
+	scmap_cell_matrix_type = [params.scmap_cell.matrix_type, null][tool_switch[params.scmap_cell.run]]
+	scpred_matrix_type = [params.scpred.matrix_type, null][tool_switch[params.scpred.run]]
+	
+	UNIQUE_MATRIX_TYPES = Channel
+	                    .of(garnett_matrix_type,
+	                        scmap_cluster_matrix_type,
+	                        scmap_cell_matrix_type,
+	                        scpred_matrix_type)
+	                    .filter{ it != null }
+	                    .unique()
+	// duplicate channel
+	UNIQUE_MATRIX_TYPES.into{ UNIQUE_MATRIX_TYPES; N_UNIQUE_MATRIX_TYPES }	
```

Comment:
Are you copying a channel into itself? Is that even legal? :)

Reply:
Ok there's a reason for this (for duplicating the channel) which I have a doubt on in. So i'm duplicating the channel to have variable with the number of unique matrices as an integer, that later on I can use [here](https://github.com/ebi-gene-expression-group/cell-types-eval-control-workflow-cross-validation/blob/74b1d000799755598215f9d6484dc8c5477cdd9b/main.nf#L160) . So instead of the `.buffer(size: 4, skip:2)` I'd have `.buffer(size: N_UNIQUE_MATRICES*2, skip:2)`. 
Do you know if this is possible?

Reply:
So, you want the contents of `N_UNIQUE_MATRICES` to be evaluated at runtime? I think this might work if you use e.g. `.first()`, but I'm not 100% sure.

Reply:
I would simply like that integer value to be used in `.buffer(size: N_UNIQUE_MATRICES*2, skip:2)`

Reply:
you can give it a try and see if it works